quic/s2n-quic-qns/src/xdp.rs

```diff
@@ -0,0 +1,304 @@
+
+        let program: &mut programs::Xdp = bpf
+            .program_mut(bpf::PROGRAM_NAME)
+            .expect("missing default program")
+            .try_into()?;
+        program.load()?;
+
+        // attach the BPF program to the interface
+        let link_id = program.attach(&interface, xdp_mode)?;
+
+        let bpf_task = async move {
+            // register the port as active
+            let mut ports: HashMap<_, _, _> = bpf
+                .map_mut(bpf::PORT_MAP_NAME)
+                .expect("missing port map")
+                .try_into()?;
+
+            ports.insert(port, 1u8, 0)?;
+
+            // register all of the RX sockets on each of the queues
+            let mut xskmap: XskMap<&mut MapData> = bpf
+                .map_mut(bpf::XSK_MAP_NAME)
+                .expect("missing socket map")
+                .try_into()?;
+
+            for (queue_id, socket) in &rx_fds {
+                xskmap.set(*queue_id, socket.as_raw_fd(), 0)?;
+            }
+
+            // print xdp stats every second, if configured
+            if xdp_stats {
+                loop {
+                    tokio::time::sleep(core::time::Duration::from_secs(1)).await;
+                    for (queue_id, socket) in &rx_fds {
+                        if let Ok(stats) = syscall::statistics(socket) {
+                            println!("stats[{queue_id}]: {stats:?}");
+                        }
+                    }
+                }
+            }
+
+            // we want this future to go until the end of the program so we can keep the BPF
+            // program active on the the NIC.
+            core::future::pending::<()>().await;
+
+            // retain the bpf program for the duration of execution
+            let _ = bpf;
+            let _ = link_id;
+            let _ = rx_fds;
+
+            Result::<(), crate::Error>::Ok(())
+        };
+
+        tokio::spawn(async move {
+            if let Err(error) = bpf_task.await {
+                panic!("BPF ERROR: {error}");
+            }
+        });
+
+        Ok(())
+    }
+
+    pub fn server(&self, addr: SocketAddr) -> Result<impl io::Provider> {
+        let socket = std::net::UdpSocket::bind(addr)?;
+        let socket = UdpSocket::from_std(socket)?;
+
+        let (umem, rx, rx_fds, tx) = self.setup()?;
+
+        self.bpf_task(addr.port(), rx_fds)?;
+
+        let io_rx = xdp_io::rx::Rx::new(rx, umem.clone());
+
+        let io_tx = {
+            // use the default encoder configuration
+            let mut encoder = encoder::Config::default();
+            if self.no_checksum {
+                encoder.set_checksum(false);
+            }
+            xdp_io::tx::Tx::new(tx, umem, encoder)
+        };
+
+        let provider = Provider::builder()
+            .with_rx(io_rx)
+            .with_tx(io_tx)
+            .with_frame_size(self.frame_size as _)?
+            .build();
+
+        tokio::spawn(async move {
+            let mut buffer = vec![0; 1500];
```

Comment:
where does the `1500` come from?

Reply:
just an arbitrary value. i'll just set it to `self.frame_size` to be consistent